sns/sns.go

```diff
@@ -0,0 +1,122 @@
+// Attribution: robbiet480's sns.go repo (https://github.com/robbiet480/go.sns) was used as the starting point for this file under the MIT License.
+
+// Package sns provides AWS SNS related functionality.
+package sns
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/go-playground/errors/v5"
+)
+
+type Payload struct {
+	Message          string `json:"Message"`
+	MessageID        string `json:"MessageId"`
+	Signature        string `json:"Signature"`
+	SignatureVersion string `json:"SignatureVersion"`
+	SigningCertURL   string `json:"SigningCertURL"`
+	SubscribeURL     string `json:"SubscribeURL"`
+	Subject          string `json:"Subject"`
+	Timestamp        string `json:"Timestamp"`
+	Token            string `json:"Token"`
+	TopicArn         string `json:"TopicArn"`
+	Type             string `json:"Type"`
+	UnsubscribeURL   string `json:"UnsubscribeURL"`
+}
+
+var hostPattern = regexp.MustCompile(`^sns\.[a-zA-Z0-9\-]{3,}\.amazonaws\.com(\.cn)?$`)
```

Comment:
**Conversation on what this regex does**:
- The host can have an optional `.cn` on the end of the domain.
- The host must end in `.amazonaws.com` or `amazonaws.com.cn`
- The host must start with `sns.`
- Between the start and end, we can have a single alpha-numeric string that can contain zero or more `-` and must be a minimum of 3 characters long.

This seems like a global solution, but I question if we should include the `.cn` in the hostname. For our usage, this seems like an obvious problem if `.cn` was found. 

There is also the consideration of this being a public repo.

Reply:
When reading AWS Docs, there is no mention of validating the host. It seems like that is an oversight...

This makes me want to re-evaluate if check is sufficient...

https://docs.aws.amazon.com/sns/latest/dg/sns-verify-signature-of-message.html

Reply:
It seems like a good "extra" check from a security standpoint instead of blindly making a request to the url in the `SigningCertURL` field.

Reply:
I'm fine with removing the .cn option if that's what you would prefer.

Reply:
@jwatson-ccc 
Let me know what you want to do to move this forward please. This is a blocker for one of my current PRs.

Reply:
The docs say to "_Verify the certificate was received from Amazon SNS_". We're doing this by validating the host.

Reply:
> It seems like a good "extra" check from a security standpoint instead of blindly making a request to the url in the `SigningCertURL` field.

Absolutely we should be doing it. It just puzzles me that this is not mentioned in AWS Documentation. This leads me to wonder what else might be missing.

Reply:
Does "_Verify the certificate was received from Amazon SNS_" qualify?